nncf/quantization/algorithms/layerwise/openvino_iterator.py

```diff
@@ -0,0 +1,232 @@
+# Copyright (c) 2024 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+from typing import Dict, List, Optional, Tuple
+
+import openvino.runtime as ov
+
+from nncf.common.graph import NNCFGraph
+from nncf.common.graph import NNCFNode
+from nncf.common.graph.transformations.layout import TransformationLayout
+from nncf.data.dataset import Dataset
+from nncf.experimental.tensor.tensor import Tensor
+from nncf.openvino.engine import OVNativeEngine
+from nncf.openvino.graph.model_transformer import OVModelTransformer
+from nncf.openvino.graph.node_utils import get_parameter_node_name
+from nncf.openvino.graph.node_utils import get_result_node_name
+from nncf.openvino.graph.transformations.commands import OVStateLessModelExtractionCommand
+from nncf.quantization.algorithms.layerwise.iterator import LayerwiseIterator
+from nncf.quantization.algorithms.layerwise.scheduler import LayerwiseStep
+from nncf.quantization.algorithms.layerwise.scheduler import NodeOutputPort
+
+
+class OVLayerwiseIterator(LayerwiseIterator):
+    """
+    OVLayerwiseIterator is a class for iterating through layers of an OpenVINO model in a layer-wise manner.
+    """
+
+    def __init__(
+        self,
+        model: ov.Model,
+        graph: NNCFGraph,
+        schedule: List[LayerwiseStep],
+        dataset: Dataset,
+        subset_size: int = 100,
+        cache: Optional[Dict[NodeOutputPort, List[Tensor]]] = None,
+    ):
+        """
+        :param model: The OpenVINO model to iterate over.
+        :param graph: The NNCF graph representation of the model.
+        :param schedule: The schedule of steps for layer-wise iteration.
+        :param dataset: The dataset to use for model inference.
+        :param subset_size: The size of the subset of the dataset to use for each iteration.
+        :param cache: The cache for storing tensors.
+        """
+        self._model = model
+        self._graph = graph
+        self._schedule = schedule
+        self._dataset = dataset
+        self._subset_size = subset_size
+        self._cache = {}  # type: Dict[NodeOutputPort, List[Tensor]]
```

Comment:
```suggestion
        self._cache: Dict[NodeOutputPort, List[Tensor]] = {}
```
Typehint by comments is old style.

Reply:
changed